### ISSUE TITLE
Illegal Access (Out of Bounds) error when strings scoped as background

### ISSUE THREAD

**Issue by nnkn** [markw65/prettier-extension-monkeyc#7]
The Error msg is :
~~~
Error: Illegal Access (Out of Bounds)
Details: Failed invoking <symbol>
Stack: 
  - makeRequest() at E:\watchface1\MyProject\bin\optimized\group003-debug\source\source\Background.mc:87 0x1000025d 
  - onTemporalEvent() at E:\watchface1\MyProject\bin\optimized\group003-debug\source\source\Background.mc:35 0x100000bd 
~~~

and the source code is 
~~~js
function makeRequest() {
...
>>>>>this line >>>>  params /*>res<*/["summary"] = Application.loadResource(Rez.Strings.failGetLocation);
...
}
~~~

the string line in xml file is : 
~~~xml
    <string scope="background" id="failGetLocation">Failed to get location</string>
~~~

when I trigger background events, the error occured.

but it runs well when I don't optimize the project  (run the project normally )

version : v2.0.54

if you need more infomation , plz tell me .

**Comment by markw65** [comment]
This could be tricky, because I've not seen anything like that. But lets start with this:

- did you enable the "Post Build Optimizer" in settings (its off by default)? If so does it still happen when you turn it off?
- if it happens with "Post Build Optimizer" turned off, does it still happen if you turn off "Minimize Modules" (which is on by default)?
- if neither of those steps help, can you verify that `params` is being set up correctly in the optimized code? It looks like the "Minimize variables" step has re-used `params` where the original code used `res`; maybe something went wrong and params is null? So actually, there's a third thing you could try: turn off "Minimize Variables".

**Comment by nnkn** [comment]
> This could be tricky, because I've not seen anything like that. But lets start with this:
> 
> * did you enable the "Post Build Optimizer" in settings (its off by default)? If so does it still happen when you turn it off?
> * if it happens with "Post Build Optimizer" turned off, does it still happen if you turn off "Minimize Modules" (which is on by default)?
> * if neither of those steps help, can you verify that `params` is being set up correctly in the optimized code? It looks like the "Minimize variables" step has re-used `params` where the original code used `res`; maybe something went wrong and params is null? So actually, there's a third thing you could try: turn off "Minimize Variables".

it works well when I turn off the "Minimize Modles" option. thank you !

**Comment by markw65** [comment]
I'd still like to try and figure out what went wrong though - `Minimize Modules` is a very useful optimization, and in theory it shouldn't break anything.

I did find one [bug in Garmin's compiler](https://forums.garmin.com/developer/connect-iq/i/bug-reports/referencing-an-imported-module-doesn-t-run-its-parent-s-init) relating to it - but my optimizer should take account of that and avoid any related issues (note that the discussion *does* mention background resources, but that's not relevant here; Jim simply didn't understand the issue).

Can you confirm that it really did call `Application.loadResource(Rez.Strings.failGetLocation);` in the optimized code, and not `Application.loadResource(Strings.failGetLocation);`?

**Comment by markw65** [comment]
Actually, I found the problem.

I've [filed a bug against Garmin's compiler](https://forums.garmin.com/developer/connect-iq/i/bug-reports/import-rez-or-using-rez-breaks-background-resources), and I'll see what I can do to avoid the issue in the optimizer.

**Comment by nnkn** [comment]
> Actually, I found the problem.
> 
> I've [filed a bug against Garmin's compiler](https://forums.garmin.com/developer/connect-iq/i/bug-reports/import-rez-or-using-rez-breaks-background-resources), and I'll see what I can do to avoid the issue in the optimizer.

wow, you are so efficient !! Wondering your new version !!

so. do I need to do this below?

> Can you confirm that it really did call Application.loadResource(Rez.Strings.failGetLocation); in the optimized code, and not Application.loadResource(Strings.failGetLocation);?

If it helps , plz tell me what to do confirm it . 
if you just asked the source code , yes, the orginal code is 
`Application.loadResource(Rez.Strings.failGetLocation);`, 
not
`Application.loadResource(Strings.failGetLocation);`

**Comment by markw65** [comment]
Thanks - I don't need anything else, I can repro the problem here.

Now I just need to figure out how to determine exactly when it's a problem and avoid it, while not killing the `Minimize Modules` optimization altogether.

**Comment by markw65** [comment]
Fixed in v2.0.55

**Comment by nnkn** [comment]
> Fixed in v2.0.55

excellent！！
